File: src/shared/protocol.js
/** Versioned client↔server message helpers. */
(function (root) {
  const PROTOCOL_VERSION = 1;

  const TYPES = {
    HELLO: "HELLO",
    WELCOME: "WELCOME",
    ROSTER: "ROSTER",
    SET_ROLE: "SET_ROLE",
    KICK: "KICK",
    SET_DURATION: "SET_DURATION",
    SET_VERSUS_GOAL: "SET_VERSUS_GOAL",
    READY: "READY",
    COLOR_CLAIM: "COLOR_CLAIM",
    MODE_CHANGE: "MODE_CHANGE",
    SETTINGS_SYNC: "SETTINGS_SYNC",
    PLAY_SYNC: "PLAY_SYNC",
    SESSION_START: "SESSION_START",
    SESSION_END: "SESSION_END",
    INPUT: "INPUT",
    STATE_DELTA: "STATE_DELTA",
    STATE_SNAPSHOT: "STATE_SNAPSHOT",
    SCORE_PULSE: "SCORE_PULSE",
    ATTEMPT_TICK: "ATTEMPT_TICK",
    ATTEMPT_EXPIRED: "ATTEMPT_EXPIRED",
    ADMIN_TRANSFER: "ADMIN_TRANSFER",
    RESYNC_REQUEST: "RESYNC_REQUEST",
    BOARD_DELTA: "BOARD_DELTA",
    BOARD_SNAPSHOT: "BOARD_SNAPSHOT",
    SPECTATE_FOCUS: "SPECTATE_FOCUS",
    /** Native co-op: per-player snake pose/colors (server relay). */
    SNAKE_DELTA: "SNAKE_DELTA",
    /** Native co-op: shared collectables from designated owner. */
    COLLECTABLES_DELTA: "COLLECTABLES_DELTA",
    /** Native co-op: player died (corpse stays); server may end when all dead. */
    COOP_PLAYER_DEAD: "COOP_PLAYER_DEAD",
    /** Native co-op: native goal hit (e.g. ALL apples). */
    COOP_GOAL: "COOP_GOAL",
    ERROR: "ERROR",
    PING: "PING",
    PONG: "PONG",
  };

  let seqCounter = 0;

  function envelope(type, payload, from) {
    return {
      v: PROTOCOL_VERSION,
      type,
      from: from || null,
      seq: ++seqCounter,
      payload: payload || {},
      ts: Date.now(),
    };
  }

  function parseMessage(raw) {
    let msg;
    try {
      msg = typeof raw === "string" ? JSON.parse(raw) : raw;
    } catch (e) {
      return { ok: false, error: "malformed_json", detail: String(e) };
    }
    if (!msg || typeof msg !== "object") {
      return { ok: false, error: "not_object" };
    }
    if (msg.v !== PROTOCOL_VERSION) {
      return { ok: false, error: "version_mismatch", got: msg.v };
    }
    if (!msg.type || typeof msg.type !== "string") {
      return { ok: false, error: "missing_type" };
    }
    return { ok: true, msg };
  }

  function encode(msg) {
    return JSON.stringify(msg);
  }

  const API = {
    PROTOCOL_VERSION,
    TYPES,
    envelope,
    parseMessage,
    encode,
    resetSeq() {
      seqCounter = 0;
    },
  };

  root.MultiplayerProtocol = API;
  if (typeof module !== "undefined" && module.exports) module.exports = API;
})(typeof window !== "undefined" ? window : globalThis);

File: server/src/protocol.rs
//! Versioned JSON protocol envelopes.

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MsgType {
    Hello,
    Welcome,
    Roster,
    SetRole,
    Kick,
    SetDuration,
    Ready,
    ColorClaim,
    ModeChange,
    SettingsSync,
    PlaySync,
    SessionStart,
    SessionEnd,
    Input,
    StateDelta,
    StateSnapshot,
    ScorePulse,
    AttemptTick,
    AttemptExpired,
    AdminTransfer,
    ResyncRequest,
    BoardDelta,
    BoardSnapshot,
    SpectateFocus,
    Error,
    Ping,
    Pong,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Envelope {
    pub v: u32,
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(default)]
    pub from: Option<String>,
    #[serde(default)]
    pub seq: u64,
    #[serde(default)]
    pub payload: Value,
    #[serde(default)]
    pub ts: Option<i64>,
}

impl Envelope {
    pub fn new(msg_type: &str, payload: Value) -> Self {
        Self {
            v: PROTOCOL_VERSION,
            msg_type: msg_type.to_string(),
            from: None,
            seq: 0,
            payload,
            ts: Some(chrono::Utc::now().timestamp_millis()),
        }
    }

    pub fn with_from(mut self, from: impl Into<String>) -> Self {
        self.from = Some(from.into());
        self
    }

    pub fn with_seq(mut self, seq: u64) -> Self {
        self.seq = seq;
        self
    }
}

pub fn parse_envelope(raw: &str) -> Result<Envelope, String> {
    let env: Envelope = serde_json::from_str(raw).map_err(|e| format!("malformed_json:{e}"))?;
    if env.v != PROTOCOL_VERSION {
        return Err(format!("version_mismatch:got={}", env.v));
    }
    if env.msg_type.is_empty() {
        return Err("missing_type".into());
    }
    Ok(env)
}

pub fn error_envelope(code: &str, message: &str) -> Envelope {
    Envelope::new(
        "ERROR",
        serde_json::json!({ "code": code, "message": message }),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_bad_version() {
        let raw = r#"{"v":99,"type":"HELLO","payload":{}}"#;
        assert!(parse_envelope(raw).unwrap_err().contains("version_mismatch"));
    }

    #[test]
    fn accepts_hello() {
        let raw = r#"{"v":1,"type":"HELLO","payload":{"roomCode":"ABCD"}}"#;
        let e = parse_envelope(raw).unwrap();
        assert_eq!(e.msg_type, "HELLO");
    }
}
